File: src/android/src/com/che/zap/crypto/ZapCrypto.kt
package com.che.zap.crypto

import com.che.zap.crypto.EncryptionUtils.HKDF_LENGTH
import com.che.zap.crypto.EncryptionUtils.KEY_LENGTH
import com.che.zap.crypto.EncryptionUtils.MAC_LENGTH
import com.che.zap.crypto.EncryptionUtils.generateHKDFBytes
import com.che.zap.crypto.EncryptionUtils.generatePublicKey
import com.che.zap.crypto.EncryptionUtils.generateSharedSecretBytes
import com.che.zap.crypto.EncryptionUtils.publicKeyToByteArray
import org.bouncycastle.crypto.engines.AESEngine
import org.bouncycastle.crypto.modes.CCMBlockCipher
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.KeyParameter
import java.nio.ByteBuffer
import java.security.interfaces.ECPublicKey

class ZapCrypto(private val localKeyProvider: LocalKeyProvider) {

    private val aesEngine = AESEngine()

    private var encryptionKeyBytes: ByteArray? = null
    private var ivBytes: ByteArray? = null
    private var counter: Int = 0

    fun initialise(devicePublicKeyBytes: ByteArray) {
        val hkdfBytes: ByteArray = generateHmacKeyDerivationFunctionBytes(devicePublicKeyBytes)
        this.encryptionKeyBytes = hkdfBytes.copyOfRange(0, KEY_LENGTH)
        this.ivBytes = hkdfBytes.copyOfRange(32, HKDF_LENGTH)
    }

    fun encrypt(data: ByteArray): ByteArray {

        assert(encryptionKeyBytes != null && ivBytes != null) { "Not initialised" }

        val counterValue = counter
        this.counter++

        val nonceBytes: ByteArray = createNonce(ivBytes!!, counterValue)
        val encrypted = encryptDecrypt(true, nonceBytes, data)
        return createCounterBytes(counterValue).plus(encrypted)
    }

    fun decrypt(counterArray: ByteArray, payload: ByteArray): ByteArray {

        assert(encryptionKeyBytes != null && ivBytes != null) { "Not initialised" }

        val nonceBytes: ByteArray = ivBytes!!.plus(counterArray)
        return encryptDecrypt(false, nonceBytes, payload)
    }

    private fun encryptDecrypt(encrypt: Boolean, nonceBytes: ByteArray, data: ByteArray): ByteArray {
        val aeadParameters = AEADParameters(KeyParameter(encryptionKeyBytes), MAC_LENGTH * 8, nonceBytes)
        val ccmBlockCipher = CCMBlockCipher(aesEngine)
        ccmBlockCipher.init(encrypt, aeadParameters)
        val processed = ByteArray(ccmBlockCipher.getOutputSize(data.size))
        ccmBlockCipher.processBytes(data, 0, data.size, processed, 0)
        ccmBlockCipher.doFinal(processed, 0)
        return processed
    }

    private fun generateHmacKeyDerivationFunctionBytes(devicePublicKeyBytes: ByteArray): ByteArray {
        val serverPublicKey: ECPublicKey = generatePublicKey(devicePublicKeyBytes, localKeyProvider.getPublicKey().params)
        val sharedSecretBytes: ByteArray = generateSharedSecretBytes(localKeyProvider.getPrivateKey(), serverPublicKey)
        val salt = publicKeyToByteArray(serverPublicKey).plus(localKeyProvider.getPublicKeyBytes())
        return generateHKDFBytes(sharedSecretBytes, salt)
    }

    private fun createNonce(iv: ByteArray, messageCounter: Int): ByteArray {
        return iv.plus(createCounterBytes(messageCounter))
    }

    private fun createCounterBytes(messageCounter: Int): ByteArray {
        return ByteBuffer.allocate(Int.SIZE_BYTES)
            .putInt(messageCounter)
            .array()
    }

}

File: src/ios/zwift_play/zapCrypto.swift
import Foundation
import CryptoKit

class ZapCrypto {
    private let localKeyProvider: LocalKeyProvider
    private var encryptionKeyBytes: Data?
    private var ivBytes: Data?
    private var counter: Int = 0

    init(localKeyProvider: LocalKeyProvider) {
        self.localKeyProvider = localKeyProvider
    }

    func initialise(devicePublicKeyBytes: Data) {
        let hkdfBytes: Data = generateHmacKeyDerivationFunctionBytes(devicePublicKeyBytes: devicePublicKeyBytes)
        self.encryptionKeyBytes = hkdfBytes.subdata(in: 0..<EncryptionUtils.KEY_LENGTH)
        self.ivBytes = hkdfBytes.subdata(in: 32..<EncryptionUtils.HKDF_LENGTH)
    }

    func encrypt(data: Data) -> Data? {
        guard let encryptionKeyBytes = encryptionKeyBytes, let ivBytes = ivBytes else {
            assertionFailure("Not initialised")
            return nil
        }

        let counterValue = counter
        self.counter += 1

        let nonceBytes: Data = createNonce(iv: ivBytes, messageCounter: counterValue)
        return encryptDecrypt(encrypt: true, nonceBytes: nonceBytes, data: data)?.prependCounter(counterValue)
    }

    func decrypt(counterArray: Data, payload: Data) -> Data? {
        guard let ivBytes = ivBytes else {
            assertionFailure("Not initialised")
            return nil
        }

        let nonceBytes: Data = ivBytes + counterArray
        return encryptDecrypt(encrypt: false, nonceBytes: nonceBytes, data: payload)
    }

    private func encryptDecrypt(encrypt: Bool, nonceBytes: Data, data: Data) -> Data? {
        guard let encryptionKey = SymmetricKey(data: encryptionKeyBytes!) else {
            return nil
        }

        let sealedBox: AES.GCM.SealedBox
        do {
            if encrypt {
                sealedBox = try AES.GCM.seal(data, using: encryptionKey, nonce: AES.GCM.Nonce(data: nonceBytes))
            } else {
                sealedBox = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: nonceBytes), ciphertext: data, tag: Data())
                return try AES.GCM.open(sealedBox, using: encryptionKey)
            }
            return sealedBox.ciphertext + sealedBox.tag
        } catch {
            print(error)
            return nil
        }
    }

    private func generateHmacKeyDerivationFunctionBytes(devicePublicKeyBytes: Data) -> Data {
        let serverPublicKey = localKeyProvider.generatePublicKey(devicePublicKeyBytes: devicePublicKeyBytes)
        let sharedSecretBytes = localKeyProvider.generateSharedSecretBytes(serverPublicKey: serverPublicKey)
        let salt = serverPublicKey.toData() + localKeyProvider.getPublicKeyBytes()
        return EncryptionUtils.generateHKDFBytes(sharedSecretBytes: sharedSecretBytes, salt: salt)
    }

    private func createNonce(iv: Data, messageCounter: Int) -> Data {
        return iv + createCounterBytes(messageCounter: messageCounter)
    }

    private func createCounterBytes(messageCounter: Int) -> Data {
        var counterValue = messageCounter.bigEndian
        return Data(bytes: &counterValue, count: MemoryLayout.size(ofValue: counterValue))
    }
}

extension Data {
    func prependCounter(_ counter: Int) -> Data {
        var counterValue = counter.bigEndian
        var data = Data(bytes: &counterValue, count: MemoryLayout.size(ofValue: counterValue))
        data.append(contentsOf: self)
        return data
    }
}
